### ISSUE TITLE
using asyncTask without useEffect

### ISSUE THREAD

**Issue by dagda1** [dai-shi/react-hooks-async#17]
I just want to check I have not missed anything when you say there should be a better abstraction from #6.

There currently is no better abstraction, I see useEffect used in asyncTask?

I sort of struggle to see the usefulness of all these hook libraries if they cannot be used in event handlers.

I really want to check that I am not missing something about why these libraries exist when a major use case is not possible.

I have seen the start example but start does not take any arguments, what if the api call is made when say an autocomplete is selected?

**Comment by dai-shi** [comment]
OK, let's open up a new discussion here.

Because my original motivation is abotrability, which almost means async call (or fetch) lives in component lifecycle (updates), I didn't think about using it in event callbacks. If put differently, we could simply call fetch in event callbacks without a library.

Now, I would like to learn your "major use case", so that I can understand what is the fundamental problem, and how a library can help for it.

In summary, your use case is probably not covered at the moment, but I am willing to extend it if it makes sense. So, please teach me what your problem is.

> what if the api call is made when say an autocomplete is selected?

I wish we had a concrete example to play with.
Can you create a codesandbox? or I can create one if I understand the expected behavior. The autocomplete is a toggle to start a typeahead example? (But, that sounds like a use case of the current library, so I may misunderstand the expected behavior.)

**Comment by dagda1** [comment]
First of all, thank you for taking the time to answer my question.  Much appreciated.

I have a lot of buttons in a view, some will cause HTTP Gets, some Deletes etc. and because they will be triggered from event handlers I cannot use hooks which must be declared in the function body.

Most of the event handlers construct URLs due to user input.

e.g.
```javascript
     <Button
        data-testid="load-button"
        disabled={businessIdInput.trim() === ''}
        id="load-button"
        onClick={() => fetch(urlJoin('base', someSetStateValue, {method: HTTP.Get})}
        primary
      >
        Load
      </Button>
```

I could have a table with a list of rows and each row would have a button to delete which would construct the URL and call HTTP Delete with that URL.

I like the way [this package](https://github.com/slorber/react-async-hook/tree/0c8084df8af83f8e426bb3094d8608217e89f8f4) looks with the `useAsyncCallback`.

So I guess you could create a high level hook that you would later call with the dynamic arguments that would in turn call fetch.

**Comment by dai-shi** [comment]
First of all, you are not only the one who suggested this kind of use cases. In addition to #6, there's a similar discussion in #8.
So, it's me having missing some use cases. I don't have a concrete idea yet, but definitely consider this.
Will get back to you, probably with some questions.

**Comment by dai-shi** [comment]
Thinking this more, I guess the feature wanted is what is proposed in #8.
I didn't accept the solution then, but now it seems the request is valid.
Basically, a) task reusability, b) passing arguments, and c) overcoming the limitation of initial render are requested (the last one is from #6).

OK, I'm not sure if I can come up with the implementation without breaking the current API, but let me try.

**Comment by dai-shi** [comment]
https://github.com/dai-shi/react-hooks-async/issues/6#issuecomment-479487865

> > Is there a reason that can’t be fixed?

> Past implementation dealt with it. However, it was a very complex implementation with useRef, and I felt like it may not work correctly in concurrent mode in some edge cases. The current implementation is much cleaner in that sense. This library is meant to use with useEffect (= useAsyncRun), and it's not a big problem with that use case.

Well, I think I now have a better understanding of how to implement it.

Hmm, but it's not easy.

**Comment by dagda1** [comment]
I like the way it is done [here](https://github.com/slorber/react-async-hook/blob/master/src/index.ts#L204).  So you create the hook and then you get an execute method that you can pass arguments to for later use in event handlers.

I've built my own abstractions on top of this.

Is this how `start` behaves?

**Comment by dai-shi** [comment]
Yes, but as I noted above there are three limitations a, b, c at this point. I’m working on it to remove/mitigate those limitations.

**Comment by dai-shi** [comment]
@dagda1 Please try this PR #18.
It should behave just like you want. (No README ready, but questions are welcome.)

**Comment by dai-shi** [comment]
Implemented in #18. Please use the latest version.
